### ISSUE TITLE
Custom Validation Type not evaluated when null

### ISSUE THREAD

**Issue by LeonardoGentile** [balderdashy/waterline#1233]
I have this sort of model

```
var Passport = {
    schema: true,

    // custom validation
    types: {
        isNonEmptyPassword: function(password) {
          if (this.protocol == 'local') {
            return password.length > 0;
          }
          return true;
        }
      },

    attributes: {
        protocol: {
            type: 'alphanumeric',
            required: true
        },
        password: {
            type: 'string',
            minLength: 8,
            maxLength: 16,
            isNonEmptyPassword: true
        },
       // ...
}
```

What I'm trying to do is: if I have a local authentication strategy check that the password is not empty with `isNonEmptyPassword`, then perform the rest of the validation (minLenght, maxLenght, and so on). 

If the authentication strategy is not local the passport is valid even tough the password is empty or null.

The problem is that the function ` isNonEmptyPassword` is never evaluated and I might have a hint: from [line 144](https://github.com/balderdashy/waterline/blob/master/lib/waterline/core/validations.js#L144) of `waterline/lib/waterline/core/validations.js`

```
    // If value is not required and empty then don't
    // try and validate it
    if(!curValidation.required) {
      if(value === null || value === '') return cb();
    }
```
In fact putting a brekpoint here I can see that the validation is not performed and the `cb()` is returned instead.

Am I wrong? Any workaround for a temp fix?

**Comment by LeonardoGentile** [comment]
I've also tried:
```
attributes: {
  protocol: {
    type: 'alphanumeric',
    required: true
  },
  password: {
    type: 'string',
    required: function maybe() {
      if (this.protocol == 'local') {
        return true;
      }
      return false;
},
```
It seems that the function is correctly evaluated, returning true or false depending on the protocol. But this doesn't work either, when my protocol is not local so that the function produce `required: false` I still receive a validation error `"required" validation rule failed for input: ''`

**Comment by sailsbot** [comment]
Thanks for posting, @LeonardoGentile.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by particlebanana** [comment]
This is a weird situation because you want two sets of rules for the value. It might be better to move the validation out into your business logic before it hits the ORM.

**Comment by LeonardoGentile** [comment]
@particlebanana yes but then I won't get back a validation error. Is there a way to throw a custom validation error even tough we skip the normal validation flow?

On a side note: can we get rid of @sailsbot ?!? I prefer to talk to humans instead of having a bot closing any issue after 30 days..

**Comment by particlebanana** [comment]
@LeonardoGentile Waterline builds validation errors like [this](https://github.com/balderdashy/waterline/blob/master/lib/waterline/query/validate.js#L47-L50) using the [WLValidationError module](https://github.com/balderdashy/waterline/blob/master/lib/waterline/error/WLValidationError.js).

Conditional validations based on other attribute values are a bit out of the scope of what Waterline validations can do.
